device/transport/amqp/src/amqp.ts

```diff
@@ -419,9 +434,19 @@ export class Amqp extends EventEmitter implements DeviceTransport {
           disableTwinDesiredPropertiesUpdates: (callback) => this._twinClient.disableTwinDesiredPropertiesUpdates(handleResult('could not disable twin desired properties updates', callback)),
           enableC2D: (callback) => {
             /*Codes_SRS_NODE_DEVICE_AMQP_41_003: [The `enableC2D` method shall attach the C2D link only if it is not already attached.] */
-            if (!this._c2dLink) {
+            if (this._c2dLink) {
+              debug('C2D link already attached, doing nothing....');
+              process.nextTick(callback);
+            } else if (this._c2dLinkAttaching) {
```

Comment:
I think you face an issue that in the time betwwen starting to attach the c2d link and it's completion

1) you get a new enable.
2) you go ahead and invoke the callback which gives the implication that c2d is enabled when it's not.
3) the actual enable then finishes 20 ticks later

Honestly, I think your solution to the send line problem was pretty darn good.  Can you use it here?  If you can, you would then be for the most part be repeating code.  Time to put enabling code into a function at that point?  Darn booleans.  Can't trust them.

Reply:
I see it. Good catch.